src/org/python/core/JavaProxyList.java

```diff
@@ -339,38 +371,17 @@ public PyObject __call__(PyObject key) {
     private static final PyBuiltinMethodNarrow listEqProxy = new ListMethod("__eq__", 1) {
         @Override
         public PyObject __call__(PyObject other) {
-            List jList = asList();
-            if (other.getType().isSubType(PyList.TYPE)) {
```

Comment:
One needs to appreciate the subtlety of the way Python binary operators give way to each other. See the note just following https://docs.python.org/2/reference/datamodel.html#object.__ror__ . Note especially the bit about subclasses.

When I look at this line, it may be trying to do that, although I expected to see it call `other._eq(this)` or some such. So wonder if the bug isn't related to that. However, if not, I'd look at the code further up the stack that calls this, as maybe mis-handling (not correctly negating) the return value, maybe when it is `null` -- see below. I haven't looked myself (yet).

Reply:
I don't think that's the intention, and maybe I should refactor this like JavaProxySet into an isPyList(other) method. I think this is just to check if we are being compared to a Python list.

On the point of how operators should give way to each other i'm not clear that should happen in the __ne__ case. https://docs.python.org/2/reference/datamodel.html#object.__ne__ just below the doc states

> when defining __eq__(), one should also define __ne__() so that the operators will behave as expected

I think thats the problem which needed to be fixed here. I did look at the code callling this and didn't see any mistakes in the logic, but another view on that would be good. I think some comments in this might help so I will add some.

Reply:
You could define `__ne__` in terms of `__eq__`, but you need to take account of a `null` return from `__eq__`, when you don't know whether it's equal or not. I'm surprised this isn't done in `PyObject` as the default implementation of `__ne__`, `__ge__` and `__le__`, but maybe there's a good reason.

The `other._eq(this)` I was looking for is dealt with in `PyObject._eq` by re-entrancy.